### ISSUE TITLE
Fix S2933: issue should not be raised for types marked with [Serializable]

### ISSUE THREAD

**Issue by ahoisl** [SonarSource/sonar-csharp#1210]
### Description

Rule S2933 checks for fields that can be made readonly. However, for types to be serializable, fields to be serialized must not be readonly. So this produces false positives for fields in types marked [Serializable] if the field is not marked with [NonSerialized].

Remark: rule S3925 should IMHO also warn a user for readonly fields, even if the class is just marked [Serializable] and does not implement ISerializable.

### Repro steps

Create a class marked as [Serializable] similar to the following:

```csharp
[Serializable]
class Person
{
    int _birthYear;  // compliant
    [NonSerialized]
    int _notSerializedValue;  // Noncompliant
    Person(int birthYear, int notSerializedValue)
    {
        _birthYear = birthYear;
        _notSerializedValue = notSerializedValue;
    }
}
```

### Expected behavior

Rule 2933 produces a warning for the "_notSerializedValue", but no warning for "_birthYear" as this field must not be readonly for deserialization to work.

### Actual behavior

Rule 2933 produces a warning for both fields "_birthYear" and "_notSerializedValue".

### Known workarounds

None.

### Related information

* SonarC# Version = 6.7.1
* Visual Studio Version = VS 2017

**Comment by Evangelink** [comment]
Hi @ahoisl,

That's indeed a False Positive on our side! Thanks for the feedback.

**Comment by ahoisl** [comment]
Thank you for fixing this issue :+1:
